presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/RemoveUnsupportedDynamicFilters.java

```diff
@@ -0,0 +1,126 @@
+ */
+public class RemoveUnsupportedDynamicFilters
+        implements PlanOptimizer
+{
+    @Override
+    public PlanNode optimize(PlanNode plan, Session session, TypeProvider types, SymbolAllocator symbolAllocator, PlanNodeIdAllocator idAllocator, WarningCollector warningCollector)
+    {
+        return SimplePlanRewriter.rewriteWith(new RemoveUnsupportedDynamicFilters.Rewriter(), plan, new HashSet<>());
+    }
+
+    private class Rewriter
+            extends SimplePlanRewriter<Set<String>>
+    {
+        @Override
+        public PlanNode visitJoin(JoinNode node, RewriteContext<Set<String>> context)
+        {
+            JoinNode join = (JoinNode) context.defaultRewrite(node, context.get());
+
+            Optional<Expression> filterOptional = join.getFilter().map(filter -> {
+                Expression modified = removeDynamicFilters(filter, context.get());
+                if (TRUE_LITERAL.equals(modified)) {
+                    return null;
+                }
+                else if (!filter.equals(modified)) {
+                    return modified;
+                }
+                return filter;
+            });
+
+            Map<String, Symbol> dynamicFilters = ImmutableMap.copyOf(filterKeys(node.getDynamicFilters(), key -> !context.get().contains(key)));
+
+            if (!filterOptional.equals(join.getFilter()) || !dynamicFilters.equals(join.getDynamicFilters())) {
+                return new JoinNode(
+                        join.getId(),
+                        join.getType(),
+                        join.getLeft(),
+                        join.getRight(),
+                        join.getCriteria(),
+                        join.getOutputSymbols(),
+                        filterOptional,
+                        join.getLeftHashSymbol(),
+                        join.getRightHashSymbol(),
+                        join.getDistributionType(),
+                        join.isSpillable(),
+                        dynamicFilters);
+            }
+            return join;
+        }
+
+        @Override
+        public PlanNode visitFilter(FilterNode node, RewriteContext<Set<String>> context)
+        {
+            FilterNode filter = (FilterNode) context.defaultRewrite(node, context.get());
+            PlanNode source = filter.getSource();
+            if (source instanceof TableScanNode) {
+                return filter;
+            }
+
+            Expression original = node.getPredicate();
+            Expression modified = removeDynamicFilters(original, context.get());
+
+            if (original.equals(modified)) {
+                return filter;
+            }
+
+            if (TRUE_LITERAL.equals(modified)) {
+                return source;
+            }
+
+            return new FilterNode(node.getId(), node.getSource(), modified);
+        }
+
+        private Expression removeDynamicFilters(Expression expression, Set<String> removedDynamicFilters)
```

Comment:
This is bottom up approach (you keep dynamic filters in join nodes) only if there is a dynamic filter consumer.
It requires context to be mutable and aggregates context from siblings subtrees.

I think better approach would be top-down:
1. When you visit JoinNode you store ids of produced dynamic filters
2. When you visit filter above scan you only keep dynamic filters that are produced by downstream join.